File: backend/app/models/meeting.py
import enum
import uuid
from datetime import date

from sqlalchemy import Date, Enum, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.mixins import CreatedAtMixin, UUIDPrimaryKeyMixin


class MeetingStatus(str, enum.Enum):
    processing = "processing"
    processed = "processed"
    failed = "failed"


class ActionItemStatus(str, enum.Enum):
    open = "open"
    in_progress = "in_progress"
    done = "done"


class DecisionStatus(str, enum.Enum):
    pending = "pending"
    decided = "decided"
    deferred = "deferred"


class Meeting(UUIDPrimaryKeyMixin, CreatedAtMixin, Base):
    __tablename__ = "meetings"

    created_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    meeting_date: Mapped[date] = mapped_column(Date, nullable=False)
    source_text: Mapped[str] = mapped_column(Text, nullable=False)
    instructions: Mapped[str | None] = mapped_column(Text, nullable=True)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[MeetingStatus] = mapped_column(
        Enum(MeetingStatus, name="meeting_status"),
        nullable=False,
        default=MeetingStatus.processing,
    )
    failure_reason: Mapped[str | None] = mapped_column(Text, nullable=True)

    action_items: Mapped[list["ActionItem"]] = relationship(
        back_populates="meeting", order_by="ActionItem.created_at", cascade="all, delete-orphan"
    )
    decisions: Mapped[list["Decision"]] = relationship(
        back_populates="meeting", order_by="Decision.created_at", cascade="all, delete-orphan"
    )
    # Read-only: EmailDraft owns the FK (source_meeting_id) and is created directly by the
    # meeting-processing job, not through this relationship — viewonly avoids needing
    # back_populates bookkeeping across the meeting/email model modules.
    email_drafts: Mapped[list["EmailDraft"]] = relationship(
        "EmailDraft",
        primaryjoin="Meeting.id == foreign(EmailDraft.source_meeting_id)",
        viewonly=True,
        order_by="EmailDraft.created_at",
    )


class ActionItem(UUIDPrimaryKeyMixin, CreatedAtMixin, Base):
    __tablename__ = "action_items"

    meeting_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("meetings.id"), nullable=False
    )
    description: Mapped[str] = mapped_column(Text, nullable=False)
    owner: Mapped[str] = mapped_column(String(255), nullable=False)
    due_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    status: Mapped[ActionItemStatus] = mapped_column(
        Enum(ActionItemStatus, name="action_item_status"),
        nullable=False,
        default=ActionItemStatus.open,
    )

    meeting: Mapped["Meeting"] = relationship(back_populates="action_items")


class Decision(UUIDPrimaryKeyMixin, CreatedAtMixin, Base):
    __tablename__ = "decisions"

    meeting_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("meetings.id"), nullable=False
    )
    description: Mapped[str] = mapped_column(Text, nullable=False)
    decided_by: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[DecisionStatus] = mapped_column(
        Enum(DecisionStatus, name="decision_status"),
        nullable=False,
        default=DecisionStatus.decided,
    )
    deadline: Mapped[date | None] = mapped_column(Date, nullable=True)

    meeting: Mapped["Meeting"] = relationship(back_populates="decisions")

File: mobile/lib/models/meeting.dart
class ActionItem {
  final String id;
  final String description;
  final String owner;
  final DateTime? dueDate;
  final String status;

  ActionItem({
    required this.id,
    required this.description,
    required this.owner,
    this.dueDate,
    required this.status,
  });

  factory ActionItem.fromJson(Map<String, dynamic> json) => ActionItem(
        id: json['id'].toString(),
        description: (json['description'] ?? '') as String,
        owner: (json['owner'] ?? '') as String,
        dueDate: json['due_date'] != null
            ? DateTime.tryParse(json['due_date'] as String)
            : null,
        status: (json['status'] ?? '') as String,
      );

  bool get isDone => status == 'done';
}

class Decision {
  final String id;
  final String description;
  final String decidedBy;
  final String status;
  final DateTime? deadline;

  Decision({
    required this.id,
    required this.description,
    required this.decidedBy,
    required this.status,
    this.deadline,
  });

  factory Decision.fromJson(Map<String, dynamic> json) => Decision(
        id: json['id'].toString(),
        description: (json['description'] ?? '') as String,
        decidedBy: (json['decided_by'] ?? '') as String,
        status: (json['status'] ?? '') as String,
        deadline: json['deadline'] != null
            ? DateTime.tryParse(json['deadline'] as String)
            : null,
      );
}

class Meeting {
  final String id;
  final String title;
  final DateTime? meetingDate;
  final String? summary;
  final String status;
  final String? failureReason;
  final List<ActionItem> actionItems;
  final List<Decision> decisions;
  final int emailDraftCount;

  Meeting({
    required this.id,
    required this.title,
    this.meetingDate,
    this.summary,
    required this.status,
    this.failureReason,
    required this.actionItems,
    required this.decisions,
    required this.emailDraftCount,
  });

  factory Meeting.fromJson(Map<String, dynamic> json) => Meeting(
        id: json['id'].toString(),
        title: (json['title'] ?? '') as String,
        meetingDate: json['meeting_date'] != null
            ? DateTime.tryParse(json['meeting_date'] as String)
            : null,
        summary: json['summary'] as String?,
        status: (json['status'] ?? '') as String,
        failureReason: json['failure_reason'] as String?,
        actionItems: (json['action_items'] as List<dynamic>? ?? const [])
            .map((e) => ActionItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        decisions: (json['decisions'] as List<dynamic>? ?? const [])
            .map((e) => Decision.fromJson(e as Map<String, dynamic>))
            .toList(),
        emailDraftCount: (json['email_drafts'] as List<dynamic>? ?? const []).length,
      );
}
